### PR TITLE
feat(gui-v2): customizable urls in shared views

### PR DESCRIPTION
**Pull request by dolsem**
## Change Summary

Adding a way to customize URL field behavior in shared views would be nice for third-party products that integrate NocoDB shared views in their UI. I am working on a NocoDB integration for [Outline](https://github.com/outline/outline) and I'd love to be able to associate an Outline document with a NocoDB item (trying to use it as a task tracker), and replace the URL with a clickable link that says something like "View Doc" and navigates to it in the same tab instead of opening a new one.

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [x] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

This PR adds a way to specify URL regexes in query parameters to the shared view, and customize behavior of URL fields that match the regex. If the custom overlay is displayed, user can press Alt to see the original URL and edit the field, assuming my PR for editing data in shared views (https://github.com/nocodb/nocodb/pull/3374) gets merged.

This also fixes what I assume is a bug that prevents users from clicking URLs in shared views (needed z-index).

## Additional information / screenshots (optional)

image

### CODE REVIEW

File: packages/nc-gui-v2/components/cell/Url.vue [hunk 1]
```diff
@@ -74,6 +100,15 @@ watch(
   </div>
 </template>
 
+<style scoped lang="scss">
+.url-overlay {
+  width: 100%;
```
Comment (wingkwong) [inline packages/nc-gui-v2/components/cell/Url.vue:105]: you may use

```scss
.url-overlay {
  @apply w-full h-full text-center !no-underline;
}
```

**Review by wingkwong — COMMENTED** [review]
I'm not quite sure how to get the custom overlay and URL regex part. Do you mind to provide some steps or record the screen for demonstration?

File: packages/nc-gui-v2/context/index.ts [hunk 2]
```diff
@@ -26,3 +26,10 @@ export const FieldsInj: InjectionKey<Ref<any[]>> = Symbol('fields-injection')
 export const ViewListInj: InjectionKey<Ref<ViewType[]>> = Symbol('view-list-injection')
 export const EditModeInj: InjectionKey<Ref<boolean>> = Symbol('edit-mode-injection')
 export const SharedViewPasswordInj: InjectionKey<Ref<string | null>> = Symbol('shared-view-password-injection')
+interface CellUrlOptions {
+  behavior?: string
+  overlay?: string
+}
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/context/index.ts:32]: Either put the interface in a separate file or put it at the top of this file but not inbetween 😄

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

**Comment by dolsem** [comment]
> I'm not quite sure how to get the custom overlay and URL regex part. Do you mind to provide some steps or record the screen for demonstration?

Oh my bad. You can set the `url_rules` query param to a URL-encoded JSON object that maps a regex in string form to URL options. For example, the following adds overlay for all Google links and makes them open in the same tab: `?url_rules=%7B%22https%3A%2F%2Fgoogle.com%2F%3F.*%22%3A%7B%22overlay%22%3A%22Visit%2BGoogle%22%2C%22behavior%22%3A%22replace%22%7D%7D`.

File: packages/nc-gui-v2/context/index.ts (hunk 2 above)
Comment (dolsem) [reply packages/nc-gui-v2/context/index.ts]: I wasn't sure where a good place to put it was, so I put it at the top of the file.

**Comment by o1lab** [comment]
@dolsem : if the idea is to use it in another app - could having an iframe will help ? We're bit confused on how this works. Please ping me on [our discord](http://discord.nocodb.com) - happy to chat.

**Comment by wingkwong** [comment]
Few more questions. 

1. What are the possible values for `behavior `? 
2. For `overlay`, is it the text that is shown in the cell, i.e. `View Doc` in your screenshot. When it's clicked, an overlay is shown with the link content?
3. disableUrlOverlay seems not configurable?
4. Does it only show in share view? Would it be better to extract the config from column option instead of route?

My Discord ID is 3278. Feel free to ping me for a chat.

**Comment by wingkwong** [comment]
Here's the summary after the clarification.

- Currently `behavior` only accept `replace` which makes the link open in the same tab instead of the new tab
- query params such as `?url_rules=%7B%22https%3A%2F%2Fgoogle.com%2F%3F.*%22%3A%7B%22overlay%22%3A%22See%2BOn%2BGoogle%22%2C%22behavior%22%3A%22replace%22%7D%7D` are added to the shared link url. Not the url in the cell. Besides, it is constructed by user side, e.g. using Outline Plugin.

File: packages/nc-gui/components/virtual-cell/Lookup.vue [hunk 3]
```diff
@@ -32,6 +32,7 @@ const lookupColumn = computed<any>(
 )
 
 provide(MetaInj, lookupTableMeta)
+provide(CellUrlDisableOverlayInj, true)
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/virtual-cell/Lookup.vue:35]: this isn't correct, should be `ref(true)`

File: packages/nc-gui/components/cell/Url.vue [hunk 4]
```diff
@@ -59,7 +72,20 @@ watch(
...
+      :target="urlOptions?.behavior === 'replace' ? undefined : '_blank'"
+      >{{ value }}
+    </nuxt-link>
+    <nuxt-link
+      v-else-if="isValid && !disableOverlay && urlOptions?.overlay"
+      class="z-3 w-full h-full text-center !no-underline hover:opacity-75"
+      :to="url"
+      :target="urlOptions?.behavior === 'replace' ? undefined : '_blank'"
+      >{{ urlOptions.overlay }}
+    </nuxt-link>
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/cell/Url.vue:88]: Formatting is broken.

**Comment by bcakmakoglu** [comment]
@dolsem I would appreciate it if you tried to move all the functionality into a composable - makes the components more readable and extracts the functionality outside for easier maintenance.

You can use [`useInjectionState`](https://github.com/nocodb/nocodb/tree/develop/packages/nc-gui/composables/useInjectionState) to create a provide/inject based composable

See [`useSidebar`](https://github.com/nocodb/nocodb/tree/develop/packages/nc-gui/composables/useSidebar) as an example for an implemenation.

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

**Comment by dolsem** [comment]
@bcakmakoglu how's this?

**Comment by wingkwong** [comment]
@dolsem Now I can see it. However, when I click the link. It doesn't open in the same tab. Instead It opens a new one. Supposedly there will be an overlay right? If possible, please attach some screenshots.

The link that I'm using is `<SHARED_URL>?url_rules=%7B%22https%3A%2F%2Fgoogle.com%2F%3F.*%22%3A%7B%22overlay%22%3A%22See%2BOn%2BGoogle%22%2C%22behavior%22%3A%22replace%22%7D%7D
`

image

With Alt Key on (or option in Mac), I can see the original value, which is fine.

**Comment by bcakmakoglu** [comment]
@dolsem No, this isn't correct yet - let me help you out and update the composable to what it should be :)

**Comment by bcakmakoglu** [comment]
@dolsem Can you add me to your fork - that way I could push a PR into your fork and point it to your branch =)

Then you can check my changes and see if that works or not - but I mostly just wanted to simplify and it and reduce some code :)

**Comment by dolsem** [comment]
@bcakmakoglu just added you. I appreciate the help. This is my first experience with Vue.

**Comment by o1lab** [comment]
@dolsem : also we missed to touch base this week. May be you can [reschedule](calendly.com/nocodb-meeting) when its convenient for you.

**Comment by dolsem** [comment]
@wingkwong the screenshot looks fine to me. Yeah it should open in the same tab, not sure what went wrong. If you hold Ctrl btw, it opens in a new tab no matter what.

**Comment by wingkwong** [comment]
@dolsem Does it work on your local? I wasn't pressing Ctrl key.

**Comment by dolsem** [comment]
@wingkwong it did work for me locally last time I checked, yeah

**Comment by o1lab** [comment]
@wingkwong : see if you could sync up / show the problem in a video. And lets sort this out and get it in as quickly as we can :) 

ps : @dolsem : nice chatting with you earlier.

**Comment by dolsem** [comment]
@wingkwong I just double checked with the exact same query param you entered and worked for me on the most up to date branch. What browser and operating system did you use? I tested with Chrome on Windows.

**Comment by wingkwong** [comment]
Oh ok now.

**Review by wingkwong — APPROVED** [review]

**Comment by o1lab** [comment]
@all-contributors please add @dolsem for code

**Pull request merged**